GenerateEverything.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE PatternGuards #-}
+
+import qualified Data.List as List
+import Control.Applicative
+import System.Environment
+import System.IO
+import System.Exit
+import System.FilePath
+import System.FilePath.Find
+
+headerFile     = "Header"
+allOutputFile  = "Everything"
+srcDirs        = ["Cubical/HITs", "Cubical/Basics"]
+
+excludedModules :: [FilePath]
+excludedModules = map toAgdaFilePath
+  [ "Cubical.HITs.HInt"
+  , "Cubical.HITs.HitInt"
+  , "Cubical.HITs.Everything"
+  , "Cubical.HITs.SmashProduct"
+  , "Cubical.Basics.Everything"
+  ] where
+
+  toAgdaFilePath :: String -> FilePath
+  toAgdaFilePath name = concat
+    [ map (\ c -> if c == '.' then '/' else c) name
+    , ".agda"
+    ]
+
+main = do
+  args <- getArgs
+  case args of
+    [] -> return ()
+    _  -> hPutStr stderr usage >> exitFailure
+
+  header <- readFileUTF8 headerFile
+  mapM_ (generateFor header) srcDirs
+
+generateFor :: String -> FilePath -> IO ()
+generateFor header srcDir = do
+  modules <- filter isLibraryModule . List.sort <$>
+               find always
+                    (extension ==? ".agda" ||? extension ==? ".lagda")
+                    srcDir
+  headers <- mapM extractHeader modules
+
+  let mkModule str = "module " ++ str ++ " where"
+  let content      = zip modules headers
+  let fileNoExt    = srcDir ++ "/" ++ allOutputFile
+
+  writeFileUTF8 (fileNoExt ++ ".agda") $
+    unlines [ header
+            , mkModule $ map (\c -> if c == '/' then '.' else c) fileNoExt
+            , format $ filter ((`notElem` excludedModules) . fst) content
+            ]
+
+-- | Usage info.
+
+usage :: String
+usage = unlines
+  [ "GenerateEverything: A utility program for Agda's cubical library."
+  , ""
+  , "Usage: runghc GenerateEverything.hs"
+  , ""
+  , "This program should be run in the base directory of a clean checkout of"
+  , "the library."
+  , ""
+  , "The program generates documentation for the library by extracting"
+  , "headers from library modules. The output is written to " ++ allOutputFile
+  , "with the file " ++ headerFile ++ " inserted verbatim at the beginning."
+  ]
+
+-- | Returns 'True' for all Agda files except for core modules.
+
+isLibraryModule :: FilePath -> Bool
+isLibraryModule f =
+  takeExtension f `elem` [".agda", ".lagda"]
+  -- && dropExtension (takeFileName f) /= "Core"
```

Comment:
Why is this commented here? The comment above seems to indicate that it shouldn't be...

Reply:
So the "Returns ..." comment above is not correct anymore?

Reply:
Yes, I'll remove that.